README.md
# TLAW - The Last API Wrapper

[![Gem Version](https://badge.fury.io/rb/tlaw.svg)](http://badge.fury.io/rb/tlaw)
[![Build Status](https://travis-ci.org/molybdenum-99/tlaw.svg?branch=master)](https://travis-ci.org/molybdenum-99/tlaw)
[![Coverage Status](https://coveralls.io/repos/molybdenum-99/tlaw/badge.svg?branch=master)](https://coveralls.io/r/molybdenum-99/tlaw?branch=master)

**TLAW** (pronounce it like "tea+love"... or whatever) is the last (and only)
API wrapper framework you'll ever need for accessing _GET-only APIs_<sup>[*](#get-only-api)</sup>
in a consistent way (think weather, search, economical indicators, geonames and so on).

## Table Of Contents

* [Features](#features)
* [Why TLAW?](#why-tlaw)
* [Usage](#usage)
  * [URLs and params description](#urls-and-params-description)
  * [Response processing](#response-processing)
    * [Flat hashes](#flat-hashes)
    * [DataTable](#datatable)
    * [Post-processing](#post-processing)
    * [All at once](#all-at-once)
  * [Documentability](#documentability)
* [Some demos](#some-demos)
* [Installation & compatibility](#installation-&-compatibility)
* [Upcoming features](#upcoming-features)
* [Get-only API](#get-only-api)
* [Current status](#current-status)
* [Links](#links)
* [Author](#author)
* [License](#license)

## Features

* **Pragmatic**: thorougly designed with tens of real world examples in mind,
  not just your textbook's "perfect orthogonal REST API";
* **Opinionated**: goal is clean and logical Ruby libary, not just mechanical
  1-by-1 endpoint-per-method wrapper for every fancy hivemind invention;
* **Easy and readable** definitions;
* **Discoverable**: once API defined in TLAW terms, you can easily investigate
  it in runtime, obtain meaningful errors like "param `foo` is missing
  while trying to access endpoint `bar`" and so on;
* **Sane metaprogramming**: allows to define entire branchy API wrapper with
  tons of pathes and endpoints in really concise manner, while creating
  _all_ corresponding classes/methods at definition time: so, at runtime
  you have no 20-level dynamic dispatching, just your usual method calls
  with clearly defined arguments and compact backtraces.

Take a look at our "model" OpenWeatherMap [wrapper](https://github.com/molybdenum-99/tlaw/blob/master/examples/open_weather_map.rb)
and [demo](https://github.com/molybdenum-99/tlaw/blob/master/examples/open_weather_map_demo.rb)
of its usage, showing how all those things work in reality.

## Why TLAW?

There are ton of small (and not-so-small) useful APIs about world around:
weather, movies, geographical features, dictionaries, world countries
statistics... Typically, when trying to use one of them from Ruby (or,
to be honest, from any programming language), you are stuck with two
options:

1. Study and use (or invent and build) some custom hand-made Wrapper
  Library™ with ton of very custom design decisions (should responses
  be just hashes, or [Hashie](https://github.com/intridea/hashie), or
  real classes for each kind of response? What are the inputs? Where should
  api key go, to global param?); or
2. Just "go commando" (sorry for the bad pun): construct URLs yourself,
  parse responses yourself, control params (or ignore the control) yourself.

TLAW tries to close this gap: provide a base for _breath-easy_ API description
which produces solid, fast and reliable wrappers.

See also a [showcase blog post](http://zverok.github.io/blog/2017-07-31-giphy-tlaw.html) with on-the-fly
API wrapper building for GIPHY.

## Usage

### URLs and params description

```ruby
class Example < TLAW::API
  define do
    base 'http://api.example.com'

    param :api_key, required: true # this would be necessary for API instance creation
    # So, the API instance would be e = Example.new(api_key: '123')
    # ...and parameter ?api_key=123 would be added to any request

    endpoint :foo # The simplest endpoint, will query "http://api.example.com/foo"
    # And then you just do e.foo and obtain result

    endpoint :bar, '/baz.json' # Path to query rewritten, will query "http://api.example.com/baz.json"
    # Method is still e.bar, though.

    # Now, for params definition:
    endpoint :movie do
      param :id
    end
    # Method call would be movie(id: '123')
    # Generated URL would be "http://api.example.com/movie?id=123"

    # When param is part of the path, you can use RFC 6570
    # URL template standard:
    endpoint :movie, '/movies/{id}'
    # That would generate method which is called like movie('123')
    # ...and call to "http://api.example.com/movies/123"

    # Now, we can stack endpoints in namespaces
    namespace :foo do # adds /foo to path
      namespace :bar, '/baz' do # optional path parameter works
        endpoint :blah # URL for call would be "http://api.example.com/foo/baz/blah"
        # And method call would be like e.foo.bar.blah(parameters)
      end

      # URL normalization works, so you can stack in namespaces even
      # things not related to them in source API, "redesigning" API on
      # the fly.
      endpoint :books, '/../books.json' # Real URL would be "http://api.example.com/books"
      # Yet method call is still namespaced like e.foo.books
    end

    # Namespaces can have their own input parameters
    namespace :foo, '/foo/{id}' do
      endpoint :bar # URL would be "http://api.example.com/foo/123/bar
      # method call would be e.foo(123).bar
    end
  end

  # ...and everything works in all possible and useful ways, just check
  # docs and demos.
end
```

See [DSL module docs](http://www.rubydoc.info/gems/tlaw/TLAW/DSL) for
full description of all features (there are few, yet very powerful).

### Response processing

TLAW is really opinionated about response processing. Main things:

1. [Hashes are "flattened"](#flat-hashes);
2. [Arrays of hashes are converted to `DataTable`s](#datatable);
3. [Post-processors for fields are easily defined](#post-processing)

#### Flat hashes

The main (and usually top-level) answer of (JSON) API is a Hash/dictionary.
TLAW takes all multilevel hashes and make them flat.

Here is an example.

Source API responds like:

```json
{
  "meta": {
    "code": "OK",
  },
  "weather": {
    "temp": 10,
    "precipitation": 138
  },
  "location": {
    "lat": 123,
    "lon": 456
  }
  ...
}
```

But TLAW response to `api.endpoint(params)` would return you a Hash looking
this way:

```json
{
  "meta.code": "OK",
  "weather.temp": 10,
  "weather.precipitation": 138,
  "location.lat": 123,
  "location.lon": 456
  ...
}
```

Reason? If you think of it and experiment with several examples, typically
with new & unexplored API you'll came up with code like:

```ruby
p response
# => 3 screens of VERY IMPORTANT RESPONSE
p response.class
# => Hash, ah, ok
p response.keys
# => ["meta", "weather", "location"], hmmm...
p response['weather']
# => stil 2.5 screens of unintelligible details
p response['weather'].class
# => Hash, ah!
p response['weather'].keys
# => and ad infinitum, real APIs are easily go 6-8 levels down
```

Now, with "opinionated" TLAW's flattening, for _any_ API you just do
the one and final `response.keys` and that's it: you see every available
data key, deep to the deepest depth.

> NB: probably, in the next versions TLAW will return some Hash descendant,
  which would also still allow you to do `response['weather']` and receive
  that "slice". Or it would not :) We are experimenting!

#### DataTable

The second main type of a (JSON) API answer, or of a part of an answer
is an array of homogenous hashes, like:

* list of data points (date - weather at that date);
* list of data objects (city id - city name - latitude - longitude);
* list of views to the data (climate model - projected temperature);
* and so on.

TLAW wraps this kind of data (array of homogenous hashes, or tables with
named columns) into `DataTable` structure, which you can think of as an
Excel spreadsheet (2d array with named columns), or loose DataFrame
pattern implementation (just like [daru](https://github.com/v0dro/daru)
or [pandas](http://pandas.pydata.org/), but seriously simpler—and much
more suited to the case).

Imagine you have an API responding something like:

```json
{
  "meta": {"count": 20},
  "data": [
    {"date": "2016-09-01", "temp": 20, "humidity": 40},
    {"date": "2016-09-02", "temp": 21, "humidity": 40},
    {"date": "2016-09-03", "temp": 16, "humidity": 36},
    ...
  ]
}
```

With TLAW, you'll see this response this way:

```ruby
pp response
{"meta.count"=>20,
 "data"=>#<TLAW::DataTable[date, temp, humidity] x 20>}
# ^ That's all. Small and easy to grasp what is what. 3 named columns,
#   20 similar rows.

d = response['data']
# => #<TLAW::DataTable[date, temp, humidity] x 20>

d.count # Array-alike
# => 20
d.first
# => {"date" => "2016-09-01", "temp" => 20, "humidity" => 40}

d.keys # Hash-alike
# => ["date", "temp", "humidity"]
d["date"]
# => ["2016-09-01", "2016-09-02", "2016-09-03" ...

# And stuff:
d.to_h
# => {"date" => [...], "temp" => [...] ....
d.to_a
# => [{"date" => ..., "temp" => ..., "humidity" => ...}, {"date" => ...

d.columns('date', 'temp') # column-wise slice
# => #<TLAW::DataTable[date, temp] x 20>
d.columns('date', 'temp').first # and so on
# => {"date" => "2016-09-01", "temp" => 20}
```

Take a look at [DataTable docs](http://www.rubydoc.info/gems/tlaw/TLAW/DataTable)
and join designing it!

#### Post-processing

When you are not happy with result representation, you can post-process
them in several ways:

```ruby
# input is entire response, block can mutate it
transform { |hash| hash['foo'] = 'bar' }

# input is entire response, and response is fully replaced with block's
# return value
transform { |hash| hash['foo'] } # Now only "foo"s value will be response

# input is value of response's key "some_key", return value of a block
# becames new value of "some_key".
transform('some_key') { |val| other_val }

# Post-processing each item, if response['foo'] is array:
transform_items('foo') {
  # mutate entire item
  transform { |item| item.delete('bar') }

  # if item is a Hash, replace its "bar" value
  transform('bar') { |val| val.to_s }
}

# More realistic examples:
transform('meta.count', &:to_i)
transform_items('daily') {
  transform('date', &Date.method(:parse))
}
transform('auxiliary_value') { nil } # Nil's will be thrown away completely
```

See full post-processing features descriptions in
[DSL module docs](http://www.rubydoc.info/gems/tlaw/TLAW/DSL).

#### All at once

All described response processing steps are performed in this order:

* parsing and initial flattening of JSON (or XML) hash;
* applying post-processors (and flatten the response after _each_ of them);
* make `DataTable`s from arrays of hashes.

### Documentability

You do it this way:

```ruby
class MyAPI < TLAW::API
  desc %Q{
    This is API, it works.
  }

  docs 'http://docs.example.com'

  namespace :ns do
    desc %Q{
      It is some interesting thing.
    }

    docs 'http://docs.example.com/ns'

    endpoint :baz do
      desc %Q{
        Should be useful.
      }

      docs 'http://docs.example.com/ns#baz'

      param :param1,
        desc: %Q{
          You don't need it, really.
        }
    end
  end
end
```

All of above is optional, but when provided, allows to investigate
things at runtime (in IRB/pry or test scripts). Again, look at
[OpenWeatherMap demo](https://github.com/molybdenum-99/tlaw/blob/master/examples/open_weather_map_demo.rb),
it shows how docs could be useful at runtime.

## Some demos

* Full-featured API wrappers:
  * OpenWeatherMap: [source API docs](http://openweathermap.org/api),
    [wrapper](https://github.com/molybdenum-99/tlaw/blob/master/examples/open_weather_map.rb),
    extensively commented & explained
    [demo code](https://github.com/molybdenum-99/tlaw/blob/master/examples/open_weather_map_demo.rb);
  * ForecastIO: [API docs](https://developer.forecast.io/docs/v2),
    [wrapper](https://github.com/molybdenum-99/tlaw/blob/master/examples/forecast_io.rb),
    [demo code](https://github.com/molybdenum-99/tlaw/blob/master/examples/forecast_io_demo.rb);
* Demos of "fire-and-forget" wrappers:
  * Urbandictionary's small and unofficial
    [API wrapper](https://github.com/molybdenum-99/tlaw/blob/master/examples/urbandictionary_demo.rb);
  * [Partial wrapper](https://github.com/molybdenum-99/tlaw/blob/master/examples/tmdb_demo.rb)
    only for some features of large [TMDB API](docs.themoviedb.apiary.io/).
    It also shows [on-the-fly updating](https://github.com/molybdenum-99/tlaw/blob/master/examples/tmdb_demo.rb#L85)
    of already existing API wrapper to add some features.

## Installation & compatibility

Just `gem install tlaw` or add it to your `Gemfile`, nothing fancy.

Required Ruby version is 2.1+, JRuby works, Rubinius seems like not.

## Upcoming features

_(in no particular order)_

* [ ] Expose Faraday options (backends, request headers);
* [ ] Request-headers based auth;
* [ ] Responses caching;
* [ ] Response headers processing DSL;
* [ ] Paging support;
* [ ] Frequency-limited API support (requests counting);
* [ ] YARD docs generation for resulting wrappers;
* [ ] More solid wrapper demos (weather sites, geonames, worldbank);
* [ ] Approaches to testing generated wrappers (just good ol' VCR should
      work, probably);
* [ ] Splat parameters.

## Get-only API

What is those "Get-only APIs" TLAW is suited for?

* It is only for _getting_ data, not changing them (though, API may use
  HTTP POST requests in reality—for example, to transfer large request
  objects);
  * It would be cool if our weather APIs could allow things like
    `POST /weather/kharkiv {sky: 'sunny', temp: '+21°C'}` in the middle
    of December, huh? But we are not leaving in the world like this.
    For now.
* It has utterly simple authentication protocol like "give us `api_key`
  param in query" (though, TLAW plans to support more complex authentication);
* It typically returns JSON answers (though, TLAW supports XML via
  awesome [crack](https://github.com/jnunemaker/crack)).

Alongside already mentioned examples (weather and so on), you can build
TLAW-backed "get-only" wrappers for bigger APIs (like Twitter), when
"gathering twits" is all you need. (Though, to be honest, TLAW's current
authorization abilities is far simpler than
[Twitter requirements](https://dev.twitter.com/oauth/application-only)).

## Current status

It is version 0.0.1. It is tested and documented, but not "tested in
battle", just invented. DSL is subject to be refined and validated,
everything could change (or broke suddenly). Tests are lot, though.

We plan to heavily utilize it for [reality](https://github.com/molybdenum-99/reality),
that would be serious evaluation of approaches and weeknesses.

## Links

* [API Docs](http://www.rubydoc.info/gems/tlaw)

## Author

[Victor Shepelev](http://zverok.github.io/)

## License

[MIT](./LICENSE.txt).


CHANGELOG.md
# The Last API Wrapper Changelog

## 0.1.0.pre (2018-12-21)

* `Namespace#children` is an `Array` now, with `Namespace#child_index` being a `Hash` (@marcandre);
* DSL now can accept strings as a Namespace/Endpoint name (@marcandre);
* Classes' `.inspect` fixed to match Ruby's conventions (@marcandre);
* Improved object tree navigation (@marcandre):
  * `APIPath#parent`, `APIPath.parent` (immediate parent class/object);
  * `APIPath#parents`, `APIPath.parents` (parent classes/objects all the way up);
  * `Namespace#traverse` (depth-first children tree enumeration).
* Lots of refactoring (better call it "rewrite", honestly), internal structure was simplified and
  decoupled, API and DSL was kept the same (I hope).
  * I am really thankful to @marcandre and @joelvh for cleaning up job they've done. Sorry, guys,
    eventually I've rewrote the thing completely :)
* Added `shared_def`/`use_def` to reuse some common parts of definitions (thanks @marcandre for
  discussion).

## 0.0.2 (2017-07-31)

* Codebase modernized and rubocopped;
* Support for redirects;
* Support for pattern-based (regexp) postprocessors.

## 0.0.1 (2016-09-19)

Initial release.


lib/tlaw/dsl.rb
# frozen_string_literal: true

module TLAW
  # This module is core of a TLAW API definition. It works like this:
  #
  # ```ruby
  # class MyAPI < TLAW::API
  #   define do # here starts what DSL does
  #     namespace :ns do
  #
  #       endpoint :es do
  #         param :param1, Integer, default: 1
  #       end
  #     end
  #   end
  # end
  # ```
  #
  # Methods of current namespace documentation describe everything you
  # can use inside `define` blocks. Actual structure of things is a bit
  # more complicated (relate to lib/tlaw/dsl.rb if you wish), but current
  # documentation structure considered to be most informative.
  #
  module DSL
    # @!method base(url)
    #   Set entire API base URL, all endpoints and namespaces pathes are calculated relative to it.
    #
    #   **Works for:** API
    #
    #   @param url [String]

    # @!method desc(text)
    #   Set description string for API object (namespace or endpoint). It can be multiline, and
    #   TLAW will automatically un-indent excessive indentations:
    #
    #   ```ruby
    #       # ...several levels of indents while you create a definition
    #       desc %Q{
    #         This is some endpoint.
    #         And it works!
    #       }
    #
    #   # ...but when you are using it...
    #   p my_api.endpoint(:endpoint).describe
    #   # This is some endpoint.
    #   # And it works!
    #   # ....
    #   ```
    #
    #   **Works for:** API, namespace, endpoint
    #
    #   @param text [String]

    # @!method docs(link)
    #   Add link to documentation as a separate line to object description. Just to be semantic :)
    #
    #   ```ruby
    #   # you do something like
    #   desc "That's my endpoint"
    #
    #   docs "http://docs.example.com/my/endpoint"
    #
    #   # ...and then somewhere...
    #   p my_api.endpoint(:endpoint).describe
    #   # That is my endpoint.
    #   #
    #   # Docs: http://docs.example.com/my/endpoint
    #   # ....
    #   ```
    #
    #   **Works for:** API, namespace, endpoint
    #
    #   @param link [String]

    # @!method param(name, type = nil, keyword: true, required: false, **opts)
    #   Defines parameter for current API (global), namespace or endpoint.
    #
    #   Param defnition defines several things:
    #
    #   * how method definition to call this namespace/endpoint would look like: whether the
    #     parameter is keyword or regular argument, whether it is required and what is default
    #     value otherwise;
    #   * how parameter is processed: converted and validated from passed value;
    #   * how param is sent to target API: how it will be called in the query string and formatted
    #     on call.
    #
    #   Note also those things about params:
    #
    #   * as described in {#namespace} and {#endpoint}, setting path template will implicitly set
    #     params. You can rewrite this on implicit param call, for ex:
    #
    #   ```ruby
    #   endpoint :foo, '/foo/{bar}'
    #   # call-sequence would be foo(bar = nil)
    #
    #   # But you can make it back keyword:
    #   endpoint :foo, '/foo/{bar}' do
    #     param :bar, keyword: true, default: 'test'
    #   end
    #   # call-sequence now is foo(bar: 'test')
    #
    #   # Or make it strictly required
    #   endpoint :foo, '/foo/{bar}/{baz}' do
    #     param :bar, required: true
    #     param :baz, keyword: true, required: true
    #   end
    #   # call-sequence now is foo(bar, baz:)
    #   ```
    #
    #   * param of outer namespace are passed to API on call to inner namespaces and endpoints, for ex:
    #
    #   ```ruby
    #   namespace :city do
    #     param :city_name
    #
    #     namespace :population do
    #       endpoint :by_year, '/year/{year}'
    #     end
    #   end
    #
    #   # real call:
    #   api.city('London').population.by_year(2015)
    #   # Will get http://api.example.com/city/year/2015?city_name=London
    #   ```
    #
    #   **Works for:** API, namespace, endpoint
    #
    #   @param name [Symbol] Parameter name
    #   @param type [Class, Symbol] Expected parameter type. Could by some class (then parameter
    #     would be checked for being instance of this class or it would be `ArgumentError`), or
    #     duck type (method name that parameter value should respond to).
    #   @param keyword [true, false] Whether the param will go as a keyword param to method definition.
    #   @param required [true, false] Whether this param is required. It will be considered on
    #     method definition.
    #   @param opts [Hash] Options
    #   @option opts [Symbol] :field What the field would be called in API query string (it would be
    #     param's name by default).
    #   @option opts [#to_proc] :format How to format this option before including into URL. By
    #    default, it is just `.to_s`.
    #   @option opts [String] :desc Params::Base description. You could do it multiline and with
    #    indents, like {#desc}.
    #   @option opts :default Default value for this param. Would be rendered in method definition
    #     and then passed to target API _(TODO: in future, there also would be "invisible" params,
    #     that are just passed to target, always the same, as well as params that aren't passed at
    #     all if user gave default value.)_
    #   @option opts [Hash, Array] :enum Whether parameter only accepts enumerated values. Two forms
    #     are accepted:
    #
    #     ```ruby
    #     # Enumerable form
    #     param :units, enum: %i[us metric britain]
    #     # parameter accepts only :us, :metric, :britain values, and passes them to target API as is
    #     # any Enumerable is OK:
    #     param :count, enum: 1..5
    #     # ^ note that means [1, 2, 3, 4, 5], not "any Numeric between 1 and 5"
    #
    #     # hash "accepted => passed" form
    #     param :compact, enum: {true => 'gzip', false => nil}
    #     # parameter accepts true or false, on true passes "compact=gzip", on false passes nothing.
    #     ```

    # @!method namespace(name, path = nil, &block)
    #   Defines new namespace or updates existing one.
    #
    #   {Namespace} has two roles:
    #
    #   * on Ruby API, defines how you access to the final endpoint, like
    #     `api.namespace1.namespace2(some_param).endpoint(...)`
    #   * on calling API, it adds its path to entire URL.
    #
    #   **NB:** If you call `namespace(:something)` and it was already defined, current definition
    #   will be added to existing one (but it can't change path of existing one, which is reasonable).
    #
    #   **Works for:** API, namespace
    #
    #   @param name [Symbol] Name of the method by which namespace would be accessible.
    #   @param path [String] Path to add to API inside this namespace.
    #     When not provided, considered to be `/<name>`. When provided, taken literally (no slashes
    #     or other symbols added). Note, that you can use `/../` in path, redesigning someone else's
    #     APIs on the fly. Also, you can use [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.txt)
    #     URL templates to mark params going straightly into URI.
    #
    #     Some examples:
    #
    #     ```ruby
    #     # assuming API base url is http://api.example.com
    #
    #     namespace :foo
    #     # method would be foo(), API URL would be http://api.example.com/foo
    #
    #     namespace :bar, '/foo/bar'
    #     # metod would be bar(), API URL http://api.example.com/foo/bar
    #
    #     namespace :baz, ''
    #     # method baz(), API URL same as base: useful for gathering into
    #     # quazi-namespace from several unrelated endpoints.
    #
    #     namespace :quux, '/foo/quux/{id}'
    #     # method quux(id = nil), API URL http://api.example.com/foo/quux/123
    #     # ...where 123 is what you've passed as id
    #     ```
    #   @param block Definition of current namespace params, and namespaces and endpoints inside
    #     current. Note that by defining params inside this block, you can change namespace's method
    #     call sequence.
    #
    #     For example:
    #
    #     ```ruby
    #     namespace :foo
    #     # call-sequence: foo()
    #
    #     namespace :foo do
    #       param :bar
    #     end
    #     # call-sequence: foo(bar: nil)
    #
    #     namespace :foo do
    #       param :bar, required: true, keyword: false
    #       param :baz, required: true
    #     end
    #     # call-sequence: foo(bar, baz:)
    #     ```
    #
    #     ...and so on. See also {#param} for understanding what you can change here.

    # @!method endpoint(name, path = nil, **opts, &block)
    #   Defines new endpoint or updates existing one.
    #
    #   {Endpoint} is the thing doing the real work: providing Ruby API method to really call target
    #   API.
    #
    #   **NB:** If you call `endpoint(:something)` and it was already defined, current definition
    #   will be added to existing one (but it can't change path of existing one, which is reasonable).
    #
    #   **Works for:** API, namespace
    #
    #   @param name [Symbol] Name of the method by which endpoint would be accessible.
    #   @param path [String] Path to call API from this endpoint.
    #     When not provided, considered to be `/<name>`. When provided, taken literally (no slashes
    #     or other symbols added). Note, that you can use `/../` in path, redesigning someone else's
    #     APIs on the fly. Also, you can use [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.txt)
    #     URL templates to mark params going straightly into URI.
    #
    #     Look at {#namespace} for examples, idea is the same.
    #
    #   @param opts [Hash] Some options, currently only `:xml`.
    #   @option opts [true, false] :xml Whether endpoint's response should be parsed as XML (JSON
    #     otherwise & by default). Parsing in this case is performed with
    #     [crack](https://github.com/jnunemaker/crack), producing the hash, to which all other rules
    #     of post-processing are applied.
    #   @param block Definition of endpoint's params and docs. Note that by defining params inside
    #     this block, you can change endpoints's method call sequence.
    #
    #     For example:
    #
    #     ```ruby
    #     endpoint :foo
    #     # call-sequence: foo()
    #
    #     endpoint :foo do
    #       param :bar
    #     end
    #     # call-sequence: foo(bar: nil)
    #
    #     endpoint :foo do
    #       param :bar, required: true, keyword: false
    #       param :baz, required: true
    #     end
    #     # call-sequence: foo(bar, baz:)
    #     ```
    #
    #     ...and so on. See also {#param} for understanding what you can change here.

    # @!method post_process(key = nil, &block)
    #   Sets post-processors for response.
    #
    #   There are also {#post_process_replace} (for replacing entire response with something else)
    #   and {#post_process_items} (for post-processing each item of sub-array).
    #
    #   Notes:
    #
    #   * You can set any number of post-processors of any kind, and they will be applied in exactly
    #     the same order they are set.
    #   * You can set post-processors in parent namespace (or for entire API), in this case
    #     post-processors of _outer_ namespace are always applied before inner ones. That allow you
    #     to define some generic parsing/rewriting on API level, then more specific key
    #     postprocessors on endpoints. But only post-processors defined BEFORE the nested object
    #     definition would be taken into account.
    #   * Hashes are flattened again after _each_ post-processor, so if for some `key` you'll
    #     return `{count: 1, continue: false}`, response hash will immediately have
    #     `{"key.count" => 1, "key.continue" => false}`. TODO: Probably it is subject to change.
    #
    #   @overload post_process(&block)
    #     Sets post-processor for whole response. Note, that in this case _return_ value of block
    #     is ignored, it is expected that your block will receive response and modify it inplace,
    #     like this:
    #
    #     ```ruby
    #     post_process do |response|
    #       response['coord'] = Geo::Coord.new(response['lat'], response['lng'])
    #     end
    #     ```
    #     If you need to replace entire response with something else, see {#post_process_replace}
    #
    #   @overload post_process(key, &block)
    #     Sets post-processor for one response key. Post-processor is called only if key exists in
    #     the response, and value by this key is replaced with post-processor's response.
    #
    #     Note, that if `block` returns `nil`, key will be removed completely.
    #
    #     Usage:
    #
    #     ```ruby
    #     post_process('date') { |val| Date.parse(val) }
    #     # or, btw, just
    #     post_process('date', &Date.method(:parse))
    #     ```
    #
    #     @param key [String]

    # @!method post_process_items(key, &block)
    #   Sets post-processors for each items of array, being at `key` (if the key is present in
    #   response, and if its value is array of hashes).
    #
    #   Inside `block` you can use {#post_process} method as described above (but all of its actions
    #   will be related only to current item of array).
    #
    #   Example:
    #
    #   Considering API response like:
    #
    #   ```json
    #   {
    #     "meta": {"count": 100},
    #     "data": [
    #       {"timestamp": "2016-05-01", "value": "10", "dummy": "foo"},
    #       {"timestamp": "2016-05-02", "value": "13", "dummy": "bar"}
    #     ]
    #   }
    #   ```
    #   ...you can define postprocessing like this:
    #
    #   ```ruby
    #   post_process_items 'data' do
    #     post_process 'timestamp', &Date.method(:parse)
    #     post_process 'value', &:to_i
    #     post_process('dummy'){nil} # will be removed
    #   end
    #   ```
    #
    #   See also {#post_process} for some generic explanation of post-processing.
    #
    #   @param key [String]

    # @!method post_process_replace(&block)
    #   Just like {#post_process} for entire response, but _replaces_ it with what block returns.
    #
    #   Real-life usage: WorldBank API typically returns responses this way:
    #
    #   ```json
    #   [
    #      {"count": 100, "page": 1},
    #      {"some_data_variable": [{}, {}, {}]}
    #   ]
    #   ```
    #   ...e.g. metadata and real response as two items in array, not two keys in hash. We can
    #   easily fix this:
    #
    #   ```ruby
    #   post_process_replace do |response|
    #     {meta: response.first, data: response.last}
    #   end
    #   ```
    #
    #   See also {#post_process} for some generic explanation of post-processing.

    # @!method shared_def(name, &block)
    #   Define reusable parts of definition, which can be utilized with {#use_def}. Common example
    #   is pagination (for one API, pagination logic for all paginated endpoints is typically the
    #   same):
    #
    #   ```ruby
    #   # at "whole API definition" level:
    #   shared_def :pagination do
    #     param :page, Integer, field: :pg, desc: "Page number, starts from 0"
    #     param :per_page, enum: (5..50), field: :per, desc: "Number of items per page"
    #   end
    #
    #   # at each particulare endpoint definition, just
    #   use_def :pagination
    #   # ...instead of repeating the param description above.
    #   ```
    #
    #   Shared definition block may contain any DSL elements.
    #
    #   **Works for:** API, namespace, endpoint

    # @!method use_def(name)
    #   Use shared definition defined earlier. See {#shared_def} for explanation and examples.
    #
    #   **Works for:** API, namespace, endpoint

    # @private
    # If there is no content in class, YARD got mad with directives.
    # See: https://github.com/lsegal/yard/issues/1207
    DUMMY = nil
  end
end

require_relative 'dsl/api_builder'
require_relative 'dsl/base_builder'
require_relative 'dsl/endpoint_builder'
require_relative 'dsl/namespace_builder'
